### PR TITLE
syncConflict flow + error handling

### LINKED ISSUE #65: Add Sync Conflict flow
Add flow for when a sync conflict occurs.

No UX reference (yet), but I assume this will just be a simple screen with the two decks side-by-side, and the option to select one copy to become the canonical copy.

### PR DESCRIPTION
**Pull request by Tyler-Brenneman**
Closes #65

- Adds a sync conflict view when sync conflicts occur.
- Sync localChanges on logout
    - if conflicts occur, ask user to logout and lose changes or resolve
- Improved error handling for sync and login

## new package:
react-native-radio-buttons (https://github.com/ArnaudRinquin/react-native-radio-buttons)
 - react-native `SegmentedControlIOS` is only available for ios, so this package replicates the functionality.
  - can style this so the buttons wont look so 'ios'

## Remaining items:
- Style sync conflict view. I am a terrible designer so I thought I would get opinions on what the view should look like and style it in a separate issue
- test ios

image

image

### CODE REVIEW

**Comment by Tyler-Brenneman** [comment]
Functionality is all done, just styling is left. Would like to hear some ideas on what the screen should look like and style it in a different issue.

**Comment by henryc132** [comment]
What happens if you click the x instead of resolving conflicts. Can you get back to conflicts later?

**Comment by Tyler-Brenneman** [comment]
Yea you can still resolve conflicts after closing by just refreshing library again

File: Cue/app/actions/library.js [hunk 1]
```diff
@@ -80,30 +96,23 @@ function syncLibrary(localChanges): ThunkAction {
 // ex: { uuid: string, action: 'edit', cards: [{action: 'edit', uuid: string, front: "fox", back: "20XX"}] } if 1 card in deck was changed
 async function syncDeck(change) : PromiseAction {
```
Comment (mattyu007) [inline Cue/app/actions/library.js:97]: We should really look into getting Flow type annotations working for this (and for `editDeck`), would make it easier to verify that we're passing in well-formed `change` objects.

File: Cue/app/tabs/MenuProfileItem.js [hunk 2]
```diff
@@ -36,20 +36,42 @@ const styles = {
...
+    if (this.props.localChanges && this.props.localChanges.length) {
+      this.props.syncLibrary(this.props.localChanges).then(failedSyncs =>{
+        if (failedSyncs && failedSyncs.length) {
+          Alert.alert(
+            'Failed to sync changes',
+            'Resolve conflicts or logout and lose local changes',
+            [
+              {text: "Logout", onPress: () => this.props.logOut()},
+              {text: "Resolve", onPress: () => this.props.navigator.push({failedSyncs})}
+            ],
```
Comment (mattyu007) [inline Cue/app/tabs/MenuProfileItem.js]: Maybe reword this as:

-----

**You have local changes which conflict with changes in the Cue cloud**

If you log out now without resolving these conflicts, you will lose all your local changes.

`Log Out Anyway` `Resolve Conflicts`

-----

Also, add `style: 'destructive'` to the logout button so that it is styled as red on iOS.

File: Cue/app/tabs/MenuProfileItem.js [hunk 3]
```diff
@@ -36,20 +36,42 @@ const styles = {
...
+        } else {
+          this.props.logOut()
+        }
+      }).catch(e => {
+        console.warn('Failed to sync changes', e)
+        Alert.alert(
+          'Could not sync local changes',
+          'Logout and lose local changes?',
+          [{text: 'Logout', onPress: () => this.props.logOut()},
+           {text: 'Cancel', style: 'cancel'}]
```
Comment (mattyu007) [inline Cue/app/tabs/MenuProfileItem.js]: Maybe:

-----
**Some changes haven’t been synced to the Cue cloud yet**

If you log out now, you will lose all your local changes.

`Cancel` `Log Out Anyway`

-----

Again, with `style: 'destructive'` on the logout button.

File: Cue/app/actions/types.js [hunk 4]
```diff
@@ -13,6 +13,7 @@ export type Action =
   | { type: 'DECK_SYNCED', serverDeck: Deck, change: {} }
   | { type: 'DECK_EDITED', change: {} }
   | { type: 'SHARE_CODE_GENERATED', uuid: string, code: string }
+	| { type: 'DECK_CONFLICT_RESOLVED', updatedDeck: Deck}
```
Comment (mattyu007) [inline Cue/app/actions/types.js]: Spacing?

File: Cue/app/api/Library.js [hunk 5]
```diff
@@ -55,7 +55,7 @@ var LibraryApi = {
   },
 
   deleteDeck(uuid: string) {
-    let endpoint = 'api/v1/library/' + uuid
+    let endpoint = '/api/v1/library/' + uuid
     console.info('deleteing deck at endpoint ' + endpoint)
```
Comment (mattyu007) [inline Cue/app/api/Library.js]: "Deleting"

File: Cue/app/api/Library.js [hunk 6]
```diff
@@ -75,7 +75,25 @@ var LibraryApi = {
...
 		return CueApi.fetch(endpoint, method, body);
-	}
+	},
+
+  overwriteDeck(deck: Deck) {
+    let endpoint = '/api/v1/deck/' + deck.uuid
+    let method = 'POST'
+    let cards = deck.cards.map(card => {return {
+      ...card,
+      uuid: card.uuid ? card.uuid : null
```
Comment (mattyu007) [inline Cue/app/api/Library.js:85]: Oh wow, I'm surprised this endpoint uses `uuid` instead of `card_id`.

File: Cue/app/api/Library.js [hunk 7]
```diff
@@ -75,7 +75,25 @@ var LibraryApi = {
...
+    }})
+    let body = JSON.stringify({
+      name: deck.name,
+      tags: deck.tags,
+      public: Boolean(deck.public), //server returns 400 without Boolean conversion
```
Comment (mattyu007) [inline Cue/app/api/Library.js:90]: Don't these `stringify` to the same thing?

File: Cue/app/common/SyncConflict.js [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    backgroundColor: 'white',
+  },
+}
+
+class SyncConflict extends React.Component {
+  props: Props
+
+  state: {
+    dataSource: ListView.dataSource,
+    deviceOrientation: string,
```
Comment (mattyu007) [inline Cue/app/common/SyncConflict.js]: Is the `deviceOrientation` state used for anything? In `LibraryListView` it's used so that the `ListView` is forced to unmount and re-mount so that the `DeckThumbnail`s are sized correctly, but I'm not sure I see why it's used here.

File: Cue/app/common/SyncConflict.js [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    conflicts: []
```
Comment (mattyu007) [inline Cue/app/common/SyncConflict.js]: Can we use a more specific Flow type annotation here like `Array<{localDeck: Deck, useServerCopy: ?boolean}>`?

File: Cue/app/common/SyncConflict.js [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+  constructor(props: Props) {
+    super(props)
+    let ds = new ListView.DataSource({
+      rowHasChanged: (r1, r2) => r1 != r2
+    })
+    let conflicts = props.failedSyncs.map(change => {
+      return {
+        localDeck: props.localDecks.find(deck=> deck.uuid==change.uuid),
+        choice: undefined,
```
Comment (mattyu007) [inline Cue/app/common/SyncConflict.js]: Should this be `useServerCopy` instead of `choice`?

File: Cue/app/common/SyncConflict.js [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+            this.props.navigator.pop()
+          })
+        }
+      }]
+    }
+  }
+
+  _setSelection = (deck: {}, useServerDeck: boolean) =>{
+    this.setState({
+      ...this.state,
```
Comment (mattyu007) [inline Cue/app/common/SyncConflict.js]: So actually I learned this recently, you don't need to unpack the current state when calling `setState` since React will automatically perform a shallow merge of the previous state.

File: Cue/app/common/SyncConflict.js [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+  //TODO: style this page
+
+type Props = {
+  failedSyncs: Array<Deck>,
+  navigator: Navigator,
+
+  // Redux
+  localDecks: Array<Deck>
+}
```
Comment (mattyu007) [inline Cue/app/common/SyncConflict.js]: Can you also add the signature of the `resolveConflict` action here to `Props`?

File: Cue/app/common/SyncConflict.js [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+function select(store) {
+  return {
+    localDecks: store.library.decks
+  }
+}
+
+function actions(dispatch) {
+  return {
+    resolveConflict: (useServerDeck, localDeck) => {
```
Comment (mattyu007) [inline Cue/app/common/SyncConflict.js]: Type annotations (for the parameters)?

File: Cue/app/tabs/account/AccountHome.js [hunk 14]
```diff
@@ -64,9 +63,35 @@ type Props = {
...
 class AccountHome extends React.Component {
+
   _logOut = () => {
-    LoginManager.logOut()
-    this.props.logOut()
+    if (this.props.localChanges && this.props.localChanges.length) {
+      this.props.syncLibrary(this.props.localChanges).then(failedSyncs =>{
+        if (failedSyncs && failedSyncs.length) {
+          Alert.alert(
+            'Failed to sync changes with server',
```
Comment (mattyu007) [inline Cue/app/tabs/account/AccountHome.js]: If you're updating the alerts on Android, update them here for iOS as well.

File: Cue/app/tabs/library/LibraryHome.android.js [hunk 15]
```diff
@@ -63,8 +63,24 @@ class LibraryHome extends React.Component {
       refreshing: true
    });
    this.props.onSyncLibrary(this.props.localChanges).then(failedSyncs =>{
-     //TODO: issue #65
-     // will need promise to return serverDeck version of the failedSync
+     if (failedSyncs && failedSyncs.length) {
+       this.props.navigator.push({failedSyncs})
+     }
+     this.setState({
+       ...this.state,
```
Comment (mattyu007) [inline Cue/app/tabs/library/LibraryHome.android.js]: Again, don't need to unpack the current state.

File: Cue/app/tabs/library/LibraryHome.android.js [hunk 16]
```diff
@@ -63,8 +63,24 @@ class LibraryHome extends React.Component {
...
+       refreshing: false
+     });
+   })
+   .catch(e => {
+     console.warn('Failed to sync changes', e)
+     this.setState({
+       ...this.state,
```
Comment (mattyu007) [inline Cue/app/tabs/library/LibraryHome.android.js]: Same here.

File: Cue/app/tabs/library/LibraryHome.android.js [hunk 17]
```diff
@@ -63,8 +63,24 @@ class LibraryHome extends React.Component {
...
+       refreshing: false
+     });
+     Alert.alert(
+       'Could not sync local changes',
```
Comment (mattyu007) [inline Cue/app/tabs/library/LibraryHome.android.js]: Maybe **Sync failed** or even **Cue cloud sync failed**?

File: Cue/app/tabs/library/LibraryHome.ios.js [hunk 18]
```diff
@@ -46,8 +46,24 @@ class LibraryHome extends React.Component {
       refreshing: true
    });
    this.props.onSyncLibrary(this.props.localChanges).then(failedSyncs =>{
-     //TODO: issue #65
-     // will need promise to return serverDeck version of the failedSync
+     if (failedSyncs && failedSyncs.length) {
+       this.props.navigator.push({failedSyncs})
+     }
+     this.setState({
+       ...this.state,
```
Comment (mattyu007) [inline Cue/app/tabs/library/LibraryHome.ios.js]: Same comments as Android.

**Review by mattyu007 — COMMENTED** [review]
Great work, just a few comments here and there. Let's open up a separate issue for styling the `SyncConflicts` view.

File: Cue/app/tabs/MenuProfileItem.js (hunk 2 above)
Comment (mattyu007) [reply Cue/app/tabs/MenuProfileItem.js]: Also, is our logout button going to be called `Log Out` or `Sign Out`? Right now we're using `Sign Out` as the user-facing word, but since internally we're calling everything `logOut`, it might be better to be consistent and change the wording in the user-facing language as well (in `AccountHome` and `MenuProfileItem`).

File: Cue/app/api/Library.js (hunk 7 above)
Comment (Tyler-Brenneman) [reply Cue/app/api/Library.js:90]: the response we get back from sever gives an int for the public property. So if this isn't changed it stays as an int in the decks store object. So when overwrite is called the public property might be an int and stringify won't give true | false

File: Cue/app/common/SyncConflict.js (hunk 8 above)
Comment (Tyler-Brenneman) [reply Cue/app/common/SyncConflict.js]: yea i just copied boilerplate code from another component. I will remove

File: Cue/app/tabs/MenuProfileItem.js (hunk 3 above)
Comment (mattyu007) [reply Cue/app/tabs/MenuProfileItem.js]: Alternatively, you could add a little extra to better communicate why we can't sync right now and what "local changes" will be lost:

-----
**Some changes haven’t been synced to the Cue cloud yet**

Can’t connect to the Cue cloud right now.

If you log out now, you will lose all the changes you made while offline.

`Cancel` `Log Out Anyway`

-----

File: Cue/app/actions/library.js (hunk 1 above)
Comment (Tyler-Brenneman) [reply Cue/app/actions/library.js:97]: I'll create an issue for this

**Comment by Tyler-Brenneman** [comment]
Updated PR for @mattyu007 comments

I'll rebase and resolve conflicts after approval.

File: Cue/app/tabs/account/AccountHome.js [hunk 19]
```diff
@@ -64,9 +63,38 @@ type Props = {
...
 class AccountHome extends React.Component {
+
   _logOut = () => {
-    LoginManager.logOut()
-    this.props.logOut()
+    if (this.props.localChanges && this.props.localChanges.length) {
+      this.props.syncLibrary(this.props.localChanges).then(failedSyncs =>{
+        if (failedSyncs && failedSyncs.length) {
+          Alert.alert(
+            'You have local changes which conflict with changes in the Cue cloud',
```
Comment (mattyu007) [inline Cue/app/tabs/account/AccountHome.js]: Title Case: `You Have Local Changes Which Conflict with Changes in the Cue Cloud`

File: Cue/app/tabs/account/AccountHome.js [hunk 20]
```diff
@@ -64,9 +63,38 @@ type Props = {
...
+            { cancelable: false }
+          )
+        } else {
+          this.props.logOut()
+        }
+      }).catch(e => {
+        console.warn('Failed to sync changes', e)
+        Alert.alert(
+          'Some changes haven’t been synced to the Cue cloud yet',
+          'Can’t connect to the Cue cloud right now.\nIf you sign out now, you will lose all the changes you made while offline.',
```
Comment (mattyu007) [inline Cue/app/tabs/account/AccountHome.js]: This needs two newlines in a row.

File: Cue/app/tabs/account/AccountHome.js [hunk 21]
```diff
@@ -106,12 +134,14 @@ class AccountHome extends React.Component {
...
   return {
     user: store.user,
+    localChanges: store.library.localChanges,
   }
 }
 
 function actions(dispatch) {
   return {
     logOut: () => dispatch(logOut()),
+    syncLibrary: (changes) => dispatch(syncLibrary(changes))
```
Comment (mattyu007) [inline Cue/app/tabs/account/AccountHome.js:148]: Missing import for `syncLibrary` action.

File: Cue/app/tabs/account/AccountHome.js [hunk 22]
```diff
@@ -64,9 +63,38 @@ type Props = {
...
-    this.props.logOut()
+    if (this.props.localChanges && this.props.localChanges.length) {
+      this.props.syncLibrary(this.props.localChanges).then(failedSyncs =>{
+        if (failedSyncs && failedSyncs.length) {
+          Alert.alert(
+            'You have local changes which conflict with changes in the Cue cloud',
+            'If you sign out now without resolving these conflicts, you will lose all your local changes',
+            [
+              {text: "Sign Out Anyway", onPress: () => this.props.logOut(), style: 'destructive'},
+              {text: "Resolve Conflicts", onPress: () => this.props.navigator.push({failedSyncs})}
```
Comment (mattyu007) [inline Cue/app/tabs/account/AccountHome.js]: `AccountHome` doesn't have a `navigator` prop yet.

File: Cue/app/CueNavigator.js [hunk 23]
```diff
@@ -107,6 +108,8 @@ var CueNavigator = React.createClass({
       return <PlayDeckView navigator={navigator} deck={route.playDeck} {...route} />
     } else if (route.preview) {
       return <DeckPreview navigator={navigator} deck={route.preview}/>
+    } else if (route.failedSyncs) {
+      return <SyncConflict navigator={navigator} failedSyncs={route.failedSyncs} />
```
Comment (mattyu007) [inline Cue/app/CueNavigator.js:112]: Minor nitpick, but since this is supposed to be a modal screen, it should be configured on iOS to use `Navigator.SceneConfigs.FloatFromBottom` with swipe-to-dismiss gesture disabled like `PlayDeckSetupView` in `configureScene`.

File: Cue/app/tabs/MenuProfileItem.js [hunk 24]
```diff
@@ -36,20 +36,43 @@ const styles = {
...
+            { cancelable: false }
+          )
+        } else {
+          this.props.logOut()
+        }
+      }).catch(e => {
+        console.warn('Failed to sync changes', e)
+        Alert.alert(
+          'Some changes haven’t been synced to the Cue cloud yet',
+          'Can’t connect to the Cue cloud right now.\nIf you sign out now, you will lose all the changes you made while offline.',
```
Comment (mattyu007) [inline Cue/app/tabs/MenuProfileItem.js]: This needs two newlines in a row.

**Review by mattyu007 — COMMENTED** [review]
Can you rebase since #140 merged the two `LibraryHome` variants?

**Comment by Tyler-Brenneman** [comment]
Updated and rebased with master

File: Cue/app/tabs/library/LibraryHome.js [hunk 25]
```diff
@@ -80,14 +80,24 @@ class LibraryHome extends React.Component {
...
    })
+   .catch(e => {
+     console.warn('Failed to sync changes', e)
+     this.setState({
+       refreshing: false
+     });
+     Alert.alert(
+       (Platform.OS === 'android' ? 'Cue cloud sync failed' : 'Cue Cloud Sync Failed'),
+       'Cue Cloud Sync Failed',
+       'Check your Internet connection and try again.'
```
Comment (mattyu007) [inline Cue/app/tabs/library/LibraryHome.js]: I'm so confused because this shouldn't work since the args are offset by one, but the alert displays correctly... Either way, remove the extra title arg.

File: Cue/app/tabs/account/AccountHome.js [hunk 26]
```diff
@@ -57,9 +58,12 @@ const styles = {
 }
 
 type Props = {
+  navigator: Navigator
```
Comment (mattyu007) [inline Cue/app/tabs/account/AccountHome.js]: Missing comma.

**Review by mattyu007 — APPROVED** [review]
Looks good. Please fix the two things and then squash.

File: Cue/app/tabs/account/AccountHome.js (hunk 26 above)
Comment (Tyler-Brenneman) [reply Cue/app/tabs/account/AccountHome.js]: I need to get my flow working again
